zio-blocks/src/main/scala/zio/blocks/schema/Optic.scala

```diff
@@ -110,82 +111,143 @@ sealed trait Lens[F[_, _], S, A] extends Optic[F, S, A] {
-
-  def apply[F[_, _], S, T, A](first: Lens[F, S, T], second: Lens[F, T, A]): Lens[F, S, A] = new LensLens(first, second)
-
-  private case class Field[F[_, _], S, A](parent: Reflect.Record[F, S], child: Term[F, S, A])
-      extends Lens[F, S, A]
-      with Leaf[F, S, A] {
+  def apply[F[_, _], S, A](parent: Reflect.Record[F, S], child: Term[F, S, A]): Lens[F, S, A] = {
     require((parent ne null) && (child ne null))
+    new LensImpl(ArraySeq(parent), ArraySeq(child))
+  }
 
-    def structure: Reflect[F, S] = parent
-
-    def focus: Reflect[F, A] = child.value
+  def apply[F[_, _], S, T, A](first: Lens[F, S, T], second: Lens[F, T, A]): Lens[F, S, A] = {
+    val u1 = first.asInstanceOf[LensImpl[F, S, A]]
+    val u2 = second.asInstanceOf[LensImpl[F, S, A]]
+    new LensImpl(u1.parents ++ u2.parents, u1.childs ++ u2.childs)
+  }
 
-    private val register: Register[A] =
-      parent.registers(parent.fields.indexWhere(_.name == child.name)).asInstanceOf[Register[A]]
+  private case class LensImpl[F[_, _], S, A](
+    parents: ArraySeq[Reflect.Record[F, S]],
+    childs: ArraySeq[Term[F, S, A]]
+  ) extends Lens[F, S, A]
+      with Leaf[F, S, A] {
+    private[this] var bindings: Array[(Deconstructor[Any], Constructor[Any], Register[Any], RegisterOffset)] = null
+
+    private[this] def init(implicit F: HasBinding[F]): Unit = {
+      val len           = parents.length
+      val bindings      = new Array[(Deconstructor[Any], Constructor[Any], Register[Any], RegisterOffset)](len)
+      var usedRegisters = RegisterOffset.Zero
+      var i             = 0
+      while (i < len) {
+        val parent        = parents(i)
+        val deconstructor = F.deconstructor(parent.recordBinding).asInstanceOf[Deconstructor[Any]]
+        val constructor   = F.constructor(parent.recordBinding).asInstanceOf[Constructor[Any]]
+        val register = parent
+          .registers(parent.fields.indexWhere {
+            val childName = childs(i).name
+            x => x.name == childName
+          })
+          .asInstanceOf[Register[Any]]
+        usedRegisters = RegisterOffset.add(usedRegisters, parent.usedRegisters)
+        bindings(i) = (deconstructor, constructor, register, usedRegisters)
+        i += 1
+      }
+      this.bindings = bindings
+    }
 
-    def get(s: S)(implicit F: HasBinding[F]): A = {
+    override def get(s: S)(implicit F: HasBinding[F]): A = {
       val registers = Registers()
-      F.deconstructor(parent.recordBinding).deconstruct(registers, RegisterOffset.Zero, s)
-      register.get(registers, RegisterOffset.Zero)
+      if (this.bindings eq null) init
+      val bindings = this.bindings
+      var offset   = RegisterOffset.Zero
+      var x: Any   = s
+      val len      = bindings.length
+      var i        = 0
+      while (i < len) {
+        val binding = bindings(i)
+        i += 1
+        binding._1.deconstruct(registers, offset, x)
```

Comment:
Except for boxing of register offset, this looks super tight! Excellent work 💪 

Note this is the line that would change to:

```suggestion
        binding._1.deconstructN(registers, n, offset, x)
```

Or similar, after `deconstructN` is introduced, which will reduce "useless copying".